app/src/logging/formatters.py

```diff
@@ -6,22 +6,54 @@
 
 See https://docs.python.org/3/library/logging.html#formatter-objects
 """
-
 import json
 import logging
-from datetime import datetime
+from datetime import date, datetime
+from decimal import Decimal
+from enum import Enum
+from typing import Any
+from uuid import UUID
 
 import src.logging.decodelog as decodelog
 
 
+def json_encoder(obj: Any) -> Any:
+    """
+    Handle conversion of various types when logs
+    are serialized into JSON. If not specified
+    will attempt to convert using str() on the object
+    """
+
+    match obj:
+        case str() | int() | float() | bool() | None:
+            return obj
+        case datetime() | date():
+            return obj.isoformat()
+        case Decimal():
+            return str(obj)
+        case Enum():
+            return obj.value
+        case UUID():
+            return str(obj)
+        case Exception():
+            return str(obj)
+        case set():
+            # The JSON library will handle
+            # iterating over and potentially calling this again
+            return list(obj)
+        case _:
+            return str(obj)
```

Comment:
How much of this code is equivalent to the default behavior? I was looking at the reference implementation of extending the JSON encoder in the docs, https://docs.python.org/3/library/json.html, and I like that in their example implementation of "ComplexEncoder" they only need to define the special case and default to [json.JSONEncoder's default method](https://docs.python.org/3/library/json.html#json.JSONEncoder.default) otherwise, so I'm wondering if we only need to implement for things like `Enum` that are custom

Reply:
Anything besides str, int, float, bool, None, and lists of those types doesn't work without this change.

I could trim it a bit as decimal/uuid/exception all would be served by the fallback case as they're the same, I just wanted to be explicit that they're handled - could be a comment.

I made a decision here that rather than fail if we didn't have a converter, we'd defer to `str()`. Since this is only for logging, and most systems that would ingest the logs would only support basic types anyways, I feel that is fine. The problem with the way the default approach works currently is that it fails silently unless you go dig through the logs.

Reply:
I see, that's a good point, could you add that as a comment (that we want to fallback to `str()` rather than failing?)

For my own understanding, in the default approach, when it fails, what does it show in the logs? Is it an empty string or does it look like some sort of exception?

Reply:
You get a very long stack trace like:
```
TypeError: Object of type UUID is not JSON serializable
Call stack:
  ...
  File "/Users/michaelchouinard/workspace/template-application-flask/app/src/api/users/user_routes.py", line 53, in user_get
    logger.info("Successfully fetched user", extra=get_user_log_params(user))
Message: 'Successfully fetched user'
Arguments: ()
```
And as far as I can tell, that log message doesn't end up being created.